File: tools/RetroFootballMac/Sources/RetroFootballMac/MatchSimulator.swift
import Foundation

struct MatchResult {
  let events: [MatchEvent]
  let homeScore: Int
  let awayScore: Int
}

enum MatchSimulator {
  private static let phases = 24
  private static let squad = 5

  static func simulate(home: Team, away: Team, seed: UInt64) -> MatchResult {
    var rng = seed
    var homeScore = 0
    var awayScore = 0
    var events: [MatchEvent] = []

    func rand() -> Double {
      rng = rng &* 1103515245 &+ 12345
      return Double(rng & 0x7fffffff) / Double(0x7fffffff)
    }

    events.append(MatchEvent(minute: 0, text: "\(home.name) vs \(away.name) — kickoff!", isGoal: false, isHalf: false))

    for phase in 0..<phases {
      let minute = min(90, Int(Double(phase + 1) / Double(phases) * 90))
      if phase == phases / 2 {
        events.append(MatchEvent(minute: minute, text: "Half time.", isGoal: false, isHalf: true))
      }

      let attack = rand() > 0.5 ? home : away
      let defend = attack.id == home.id ? away : home
      let atk = power(attack, sw: 0.45, pw: 0.35, dw: 0.2)
      let def = power(defend, sw: 0.15, pw: 0.25, dw: 0.6)
      let roll = rand() * (atk + def)
      let shooter = pick(attack, shoot: true)

      if roll < atk * 0.12 {
        if attack.id == home.id { homeScore += 1 } else { awayScore += 1 }
        events.append(MatchEvent(
          minute: minute,
          text: "GOAL! \(shooter.name) (\(attack.name)) \(homeScore)-\(awayScore)",
          isGoal: true,
          isHalf: false
        ))
      } else if roll < atk * 0.35 {
        events.append(MatchEvent(minute: minute, text: "\(shooter.name) shoots — saved!", isGoal: false, isHalf: false))
      } else if roll < atk * 0.55 {
        events.append(MatchEvent(minute: minute, text: "\(defend.name) win the ball.", isGoal: false, isHalf: false))
      } else {
        events.append(MatchEvent(minute: minute, text: "\(shooter.name) — shot wide.", isGoal: false, isHalf: false))
      }
    }

    events.append(MatchEvent(
      minute: 90,
      text: "Full time: \(home.name) \(homeScore) - \(awayScore) \(away.name)",
      isGoal: true,
      isHalf: false
    ))

    return MatchResult(events: events, homeScore: homeScore, awayScore: awayScore)
  }

  private static func power(_ team: Team, sw: Double, pw: Double, dw: Double) -> Double {
    let squad = Array(team.players.prefix(squad))
    guard !squad.isEmpty else { return 50 }
    return squad.reduce(0.0) { $0 + Double($1.shoot) * sw + Double($1.pass) * pw + Double($1.defend) * dw + Double($1.pace) * 0.1 } / Double(squad.count)
  }

  private static func pick(_ team: Team, shoot: Bool) -> Player {
    let squad = Array(team.players.prefix(squad))
    return squad.sorted { shoot ? $0.shoot > $1.shoot : $0.pass > $1.pass }.first!
  }
}

File: Assets/Scripts/Match/MatchSimulator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using RetroFootball76.Data;

namespace RetroFootball76.Match
{
    /// <summary>
    /// Stat-driven 5-a-side arcade simulator using 1974/76 squad ratings.
    /// </summary>
    public class MatchSimulator
    {
        readonly Random _rng;
        readonly MatchState _state;

        public MatchState State => _state;

        public MatchSimulator(TeamData home, TeamData away, int seed = 0)
        {
            _rng = seed == 0 ? new Random() : new Random(seed);
            _state = new MatchState
            {
                homeTeam = home,
                awayTeam = away,
                minute = 0,
                homeScore = 0,
                awayScore = 0
            };
            AddEvent(0, MatchEventType.Kickoff, home.id, null,
                $"{home.name} vs {away.name} — kickoff!");
        }

        public void SimulateFullMatch()
        {
            var totalPhases = Core.GameConstants.PhasesPerHalf * 2;
            var minutesPerPhase = (Core.GameConstants.HalfMinutes * 2 * 60f) / totalPhases;

            for (var phase = 0; phase < totalPhases; phase++)
            {
                _state.minute = (int)Math.Min(90, (phase + 1) * minutesPerPhase / 60f * 90f / (Core.GameConstants.HalfMinutes * 2));
                if (phase == Core.GameConstants.PhasesPerHalf)
                    AddEvent(_state.minute, MatchEventType.HalfTime, null, null, "Half time.");

                var homeAttacks = _rng.NextDouble() > 0.5;
                SimulatePhase(homeAttacks ? _state.homeTeam : _state.awayTeam,
                              homeAttacks ? _state.awayTeam : _state.homeTeam);
            }

            _state.isFinished = true;
            AddEvent(90, MatchEventType.FullTime, null, null,
                $"Full time: {_state.homeTeam.name} {_state.homeScore} - {_state.awayScore} {_state.awayTeam.name}");
        }

        void SimulatePhase(TeamData attack, TeamData defend)
        {
            var atk = SquadPower(attack, shootWeight: 0.45f, passWeight: 0.35f, defendWeight: 0.2f);
            var def = SquadPower(defend, shootWeight: 0.15f, passWeight: 0.25f, defendWeight: 0.6f);
            var roll = _rng.NextDouble() * (atk + def);
            var shooter = PickPlayer(attack, preferShoot: true);

            if (roll < atk * 0.12)
            {
                Score(attack, shooter);
            }
            else if (roll < atk * 0.35)
            {
                AddEvent(_state.minute, MatchEventType.Shot, attack.id, shooter?.name,
                    $"{shooter?.name} shoots — saved!");
                AddEvent(_state.minute, MatchEventType.Save, defend.id, PickPlayer(defend, preferShoot: false)?.name,
                    "Great save!");
            }
            else if (roll < atk * 0.55)
            {
                AddEvent(_state.minute, MatchEventType.Turnover, defend.id, null, $"{defend.name} win the ball.");
            }
            else
            {
                AddEvent(_state.minute, MatchEventType.Shot, attack.id, shooter?.name,
                    $"{shooter?.name} — shot wide.");
            }
        }

        void Score(TeamData team, PlayerData scorer)
        {
            if (team.id == _state.homeTeam.id) _state.homeScore++;
            else _state.awayScore++;

            AddEvent(_state.minute, MatchEventType.Goal, team.id, scorer?.name,
                $"GOAL! {scorer?.name} ({team.name}) {_state.homeScore}-{_state.awayScore}");
        }

        static float SquadPower(TeamData team, float shootWeight, float passWeight, float defendWeight)
        {
            var squad = team.players.Take(Core.GameConstants.SquadSize).ToList();
            if (squad.Count == 0) return 50f;
            float sum = 0;
            foreach (var p in squad)
                sum += p.shoot * shootWeight + p.pass * passWeight + p.defend * defendWeight + p.pace * 0.1f;
            return sum / squad.Count;
        }

        static PlayerData PickPlayer(TeamData team, bool preferShoot)
        {
            var squad = team.players.Take(Core.GameConstants.SquadSize).ToList();
            if (squad.Count == 0) return null;
            squad.Sort((a, b) => (preferShoot ? b.shoot - a.shoot : b.pass - a.pass));
            return squad[0];
        }

        void AddEvent(int minute, MatchEventType type, string teamId, string player, string desc)
        {
            _state.events.Add(new MatchEvent
            {
                minute = minute,
                type = type,
                teamId = teamId,
                playerName = player,
                description = desc
            });
        }
    }
}
